### PR TITLE
[JENKINS-58654] loadClass is used so many times so build a cache around

### PR DESCRIPTION
**Pull request by olamy**
Signed-off-by: olivier lamy <olamy@apache.org>

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLContext.java [hunk 1]
```diff
@@ -30,6 +35,15 @@
      * The current thread's context.
      */
     private static ThreadLocal<Stack<DisplayURLContext>> context = new ThreadLocal<>();
+
+    private static final Cache<String, Optional<PluginWrapper>> CACHE =
+        CacheBuilder.newBuilder()
+            .maximumSize(Integer.getInteger( DisplayURLContext.class.getName() + ".cache.size", 1000))
+            // cache ttl default 5 minutes
+            .expireAfterAccess( Integer.getInteger( DisplayURLContext.class.getName() + ".cache.ttl", 300000), TimeUnit.MILLISECONDS)
+            //.weakValues() not sure we need this as entries are pluginWrapper will never be garbaged
```
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLContext.java]: Yeah I think you're right, it looks like the lifetime for a `PluginWrapper` is the same as the lifetime of `PluginManager`, so `weakValues` seems pointless.

**Comment by dwnusbaum** [comment]
@olamy Looking at this a bit more, if the concern here is Pipeline is [`EnvironmentContributorImpl`](https://github.com/jenkinsci/display-url-api-plugin/blob/f5fe585a93881d2c15130cdb9fef06273bd53fd1/src/main/java/org/jenkinsci/plugins/displayurlapi/EnvironmentContributorImpl.java), maybe it would be better to provide a new overload of `DisplayURLContext.open()` that take a `boolean` `guessPlugin`, and then we start using that overload [here](https://github.com/jenkinsci/display-url-api-plugin/blob/f5fe585a93881d2c15130cdb9fef06273bd53fd1/src/main/java/org/jenkinsci/plugins/displayurlapi/EnvironmentContributorImpl.java#L18) and [here](https://github.com/jenkinsci/display-url-api-plugin/blob/f5fe585a93881d2c15130cdb9fef06273bd53fd1/src/main/java/org/jenkinsci/plugins/displayurlapi/EnvironmentContributorImpl.java#L29) and then we can delete the calls that explicitly unset `plugin` [here](https://github.com/jenkinsci/display-url-api-plugin/blob/f5fe585a93881d2c15130cdb9fef06273bd53fd1/src/main/java/org/jenkinsci/plugins/displayurlapi/EnvironmentContributorImpl.java#L20) and [here](https://github.com/jenkinsci/display-url-api-plugin/blob/f5fe585a93881d2c15130cdb9fef06273bd53fd1/src/main/java/org/jenkinsci/plugins/displayurlapi/EnvironmentContributorImpl.java#L31). 

Right now the code generates a stack trace and loads classes looking for a plugin only to unset the  computed plugin later, which seems very wasteful. Probably fine to keep the cache in for other use cases, although I don't have a sense of how often that code will run.

**Comment by olamy** [comment]
@dwnusbaum yes my concern is really EnvironmentContributorImpl and your idea seems to be a good approach. I will test that as well. Thanks!

**Comment by olamy** [comment]
@dwnusbaum  well that's endless if we go this way... :( because this https://github.com/jenkinsci/display-url-api-plugin/blob/f5fe585a93881d2c15130cdb9fef06273bd53fd1/src/main/java/org/jenkinsci/plugins/displayurlapi/EnvironmentContributorImpl.java#L22 is using the open method as well but no need of the guessPlugin neither... 
Anyway I'm happy to change with your suggestion but keep the cache as well.

**Comment by olamy** [comment]
@dwnusbaum  I cannot find any direct usage of classes from package org.jenkinsci.plugins.displayurlapi in BlueOcean...

File: src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLContext.java [hunk 2]
```diff
@@ -95,39 +109,61 @@ private void guessPlugin() {
...
-                // ignore, it's not a plugin
             }
         }
     }
 
+
     /**
      * Opens a {@link DisplayURLContext} for the current thread.
-     *
+     * @param guessPlugin to try and infer the current plugin (a bit resources consuming as inspect stacktrace and classloader)
```
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLContext.java]: nit: maybe worth expanding on the comment like this:

```suggestion
     * @param guessPlugin try to infer the current plugin (resource intensive as requires a stack trace and class loading). Use {@code false} if you know the caller will always be Jenkins core.
```

File: src/main/java/org/jenkinsci/plugins/displayurlapi/EnvironmentContributorImpl.java [hunk 3]
```diff
@@ -15,9 +15,8 @@
 
     @Override
     public void buildEnvironmentFor(@NonNull Run r, @NonNull EnvVars envs, @NonNull TaskListener listener) throws IOException, InterruptedException {
-        try (DisplayURLContext ctx = DisplayURLContext.open()) {
+        try (DisplayURLContext ctx = DisplayURLContext.open(false)) {
```
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/EnvironmentContributorImpl.java]: nit: maybe worth leaving the old comment around for clarity:

```suggestion
        try (DisplayURLContext ctx = DisplayURLContext.open(false)) { // environment contributor "comes from" core
```

**Comment by dwnusbaum** [comment]
Closing and reopening for a fresh build:

> win2012-b66f50 was marked offline: Connection was broken: java.util.concurrent.TimeoutException: Ping started at 1564466541837 hasn't completed by 1564466781839

**Pull request merged**